### PR TITLE
feat/primekg loader

### LINKED ISSUE #32: Example biomedical knowledge graph for training and benchmarking
# Description
As a researcher, I would like an open-source biomedical knowledge graph that can be used for testing and benchmarking. A popular open-source knowledge graph that has been used in various benchmarking tasks is PrimeKG https://github.com/mims-harvard/PrimeKG which has been used in STARK https://github.com/snap-stanford/stark and BioBridge https://github.com/RyanWangZf/BioBridge. In particular, we are most interested in the question and answering tasks (node classification) from STARK and potentially the cross-modality prediction from BioBridge

# Checklist
- [ ] Scripts for downloading PrimeKG
- [ ] Scripts for setting up and running the STARK benchmark
- [ ] (Optional) Scripts for setting up and running the BioBridge benchmark

# Acceptance criteria
- [ ] Docs, Lints, and Unit tests as appropriate to support the scripts for downloading the data assets and running the benchmarks
- [ ] Jupyter notebook demonstrating how to download the datasets and run the benchmarks
- [ ] Successful PR

# Resources
Please see "Downloading datasets and models (Stark example)" from https://virtualpatientengine.github.io/VPEHackathonAIAgentsCO/setup/

### PR DESCRIPTION
**Pull request by awmulyadi**
# For authors

## Description

Please:
I have developed a set of codes for loading a series of biomedical knowledge graph-based datasets, i.e., PrimeKG, StarkQA-PrimeKG, and BioBridge-PrimeKG. Following are the detailed key updates:
1. **Source Code**: Within the folder of `aiagents4pharma/talk2knowledgegraphs/datasets`, I included an abstract class of `Dataset` followed by a set of implementation classes of `PrimeKG`, `StarkQAPrimeKG`, and `BioBridgePrimeKG` for loading the corresponding datasets.
2. **Pytest**: I included test cases of the above classes under the folder of `aiagents4pharma/talk2knowledgegraphs/tests/*`
3. **Tutorial Notebooks**: I added an interactive notebook for showcasing the use cases of the classes under `docs/notebooks/talk2knowledgegraphs/*`.
4. **Documentation**: Finally, I updated the related documentation using mkdocs, which is available in `docs/talk2knowledgegraphs/*`.

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test(s) in the ```tests``` folder
- [x] Added new function(s) to an existing test(s) (i.e., `aiagents4pharma/talk2knowledgegraphs/tests/*`)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. MkDocs)
- [x] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

File: aiagents4pharma/talk2knowledgegraphs/datasets/biobridge_primekg.py [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+            df_node_test.to_csv(os.path.join(self.local_dir, "processed", "node_test.tsv.gz"),
+                                sep="\t", compression="gzip", index=False)
+            # Store altered full triplets as well
+            triplets.to_csv(os.path.join(self.local_dir, "processed",
+                                         "triplet_full_altered.tsv.gz"),
+                            sep="\t", compression="gzip", index=False)
+
+        return df_train, df_node_train, df_test, df_node_test, triplets
+
+    # def _negative_sampling(self,
```
[the remark this answers is not in the document]
Comment (awmulyadi) [reply aiagents4pharma/talk2knowledgegraphs/datasets/biobridge_primekg.py:344]: Correct, I have tested this sampling function to build negative triplets for each positive one. However, it requires some time to finish the process since the training records have about 3 million records. Therefore, I decided to ignore this functionality as of now. I will look back into this one once the training requirement needs it, if it is ok with you.

File: aiagents4pharma/talk2knowledgegraphs/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+name = "aiagents4pharma-talk2knowledgegraphs"
+description = "AI Agents for drug discovery, drug development, and other pharmaceutical R&D"
+readme = "README.md"
+requires-python = ">=3.10"
+classifiers = [
+    "Programming Language :: Python :: 3",
+    "License :: OSI Approved :: MIT License",
+    "Operating System :: OS Independent",
+]
+dependencies = [
```
[the remark this answers is not in the document]
Comment (awmulyadi) [reply aiagents4pharma/talk2knowledgegraphs/pyproject.toml:15]: Yes, I agree. I took this approach since there is a library called `tiledbsoma` that is not supported for certain OS (Windows). It is kind of a chained effect: `stark` requires `PyTDC`, with `PyTDC` requiring `tiledsoma`. Hopefully, they solve the issue anytime soon.

Please check:
[https://github.com/single-cell-data/TileDB-SOMA/issues?q=is%3Aissue+windows](https://github.com/single-cell-data/TileDB-SOMA/issues?q=is%3Aissue+windows)

File: aiagents4pharma/talk2knowledgegraphs/tests/test_biobridge_primekg_loader.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    assert biobridge_triplets is not None
+    assert len(biobridge_triplets) > 0
+    assert biobridge_triplets.shape[0] == 3904610
+    assert list(biobridge_splits.keys()) == ['train', 'node_train', 'test', 'node_test']
+    assert len(biobridge_splits['train']) == 3510930
+    assert len(biobridge_splits['node_train']) == 76486
+    assert len(biobridge_splits['test']) == 393680
+    assert len(biobridge_splits['node_test']) == 8495
+
+def test_load_existing_primekg(biobridge_primekg):
```
[the remark this answers is not in the document]
Comment (awmulyadi) [reply aiagents4pharma/talk2knowledgegraphs/tests/test_biobridge_primekg_loader.py:108]: I included additional test cases for achieving full coverage tests for other methods as well.

File: docs/notebooks/talk2knowledgegraphs/tutorial_starkqa_primekg_loader.ipynb
[the remark this answers is not in the document]
Comment (awmulyadi) [reply docs/notebooks/talk2knowledgegraphs/tutorial_starkqa_primekg_loader.ipynb:1]: I have rechecked the repository of Stark about this, and apparently they provided the query and node embeddings using 'text-embedding-ada-002'. However, there is no information on edge embeddings. I have updated the class of StarkQAPrimeKG by including pre-loaded embeddings of queries and nodes. However, we also need to embed the edges (~8M) using the 'text-embedding-ada-002' in the next step (KG construction). Alternatively, I am preparing codes to embed both nodes and edges using Ollama's nomic-embed-text to be included in the next PR.

Ref:
[https://github.com/snap-stanford/stark/blob/main/emb_download.py](https://github.com/snap-stanford/stark/blob/main/emb_download.py)

[the remark this answers is not in the document]
Comment (dmccloskey) [reply docs/notebooks/talk2knowledgegraphs/tutorial_starkqa_primekg_loader.ipynb:1]: I see. I do not believe they used edge embeddings in their manuscript (?).

I wouldn't worry about embedding the edges using `text-embedding-ada-002` if time is of the essence. It does not seem that it would help us reproduce their work. Please correct me if I am wrong.

File: aiagents4pharma/talk2knowledgegraphs/tests/test_biobridge_primekg_loader.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmccloskey) [reply aiagents4pharma/talk2knowledgegraphs/tests/test_biobridge_primekg_loader.py:123]: OK cool 😊

File: aiagents4pharma/talk2knowledgegraphs/datasets/starkqa_primekg.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                [qa_indices.index(query_id) for query_id in query_ids]
+            )
+
+        # Load the node info of PrimeKG preprocessed for StarkQA
+        with open(os.path.join(self.local_dir, 'skb/prime/processed/node_info.pkl'), 'rb') as f:
+            starkqa_node_info = pickle.load(f)
+
+        return starkqa, starkqa_split_idx, starkqa_node_info
+
+    def _load_stark_embeddings(self) -> tuple:
```
Comment (dmccloskey) [inline aiagents4pharma/talk2knowledgegraphs/datasets/starkqa_primekg.py:107]: 👍

**Review by dmccloskey — APPROVED** [review]
Ready to be merged whenever you are ready.

File: docs/notebooks/talk2knowledgegraphs/tutorial_starkqa_primekg_loader.ipynb
[the remark this answers is not in the document]
Comment (awmulyadi) [reply docs/notebooks/talk2knowledgegraphs/tutorial_starkqa_primekg_loader.ipynb:1]: Correct, in their manuscript and code repository, they didn't mention edge/relation embeddings. Thus, the methods used in their benchmark most likely didn't incorporate these features.
